src/scan-imports.ts

```diff
@@ -1,159 +1,131 @@
+  if (webModulesIndex === -1) {
+    return null;
   }
-  return undefined;
-}
-
-/**
- * Return dynamic import name (or undefined)
- */
-function dynamicImportModuleName(node: NodePath<CallExpression>): string | undefined {
-  if (isCallExpression(node)) {
-    if (node.node.callee.type === 'Import' && node.node.arguments && node.node.arguments.length) {
-      return (node.node.arguments[0] as any).value;
-    }
+  // check if the resolved specifier (including file extension) is a known package.
+  const resolvedSpecifier = specifier.substring(webModulesIndex + 'web_modules/'.length);
+  if (knownDependencies.includes(resolvedSpecifier)) {
+    return resolvedSpecifier;
   }
-  return undefined;
-}
-
-/**
- * Return array of dependencies
- */
-function parseImports(file: string) {
-  const code = fs.readFileSync(file, 'utf8');
-
-  // skip file if empty, or if permissions error
-  if (!code) {
-    return [];
+  // check if the resolved specifier (without extension) is a known package.
+  const resolvedSpecifierWithoutExtension = stripJsExtension(resolvedSpecifier);
+  if (knownDependencies.includes(resolvedSpecifierWithoutExtension)) {
+    return resolvedSpecifierWithoutExtension;
   }
-  const deps = new Set<string>();
+  // Otherwise, this is an explicit import to a file within a package.
+  return resolvedSpecifier;
+}
 
-  // try to parse; this will fail if not JS
+function getInstallTargetsForFile(code: string, knownDependencies: string[]): InstallTarget[] {
+  const allImports: InstallTarget[] = [];
   try {
     const ast = parse(code, {plugins: ['dynamicImport'], sourceType: 'module'});
     traverse(ast, {
-      enter(node) {
-        const moduleName =
-          importModuleName(node as NodePath<ImportDeclaration>) ||
-          dynamicImportModuleName(node as NodePath<CallExpression>);
-
-        if (moduleName) {
-          deps.add(moduleName);
+      ImportDeclaration(path) {
+        const webModuleSpecifier = parseWebModuleSpecifier(
+          path.node.source.value,
+          knownDependencies,
+        );
+        if (webModuleSpecifier) {
+          allImports.push({
+            specifier: webModuleSpecifier,
+            all: false,
+            default: path.node.specifiers.some(s => s.type === 'ImportDefaultSpecifier'),
+            namespace: path.node.specifiers.some(s => s.type === 'ImportNamespaceSpecifier'),
+            named: path.node.specifiers
+              .map(s => s.type === 'ImportSpecifier' && s.imported.name)
+              .filter(Boolean),
+          });
+        }
+      },
+      Import(path) {
+        // Only match dynamic imports that are called as a function
+        if (path.parent.type !== 'CallExpression') {
+          return;
+        }
+        // Only match dynamic imports called with a single string argument
+        const [argNode] = path.parent.arguments;
+        if (argNode.type !== 'StringLiteral') {
+          return;
+        }
```

Comment:
Good catch! Didn’t even think of dynamic imports with references.

Reply:
np! I actually just grabbed this code from https://github.com/pikapkg/standard-pkg, which had this logic already well fleshed out & baked.